lib/yast/tarball_servlet.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+#--
+# Yast rake
+#
+# Copyright (C) 2020, SUSE LLC
+#   This library is free software; you can redistribute it and/or modify
+# it only under the terms of version 2.1 of the GNU Lesser General Public
+# License as published by the Free Software Foundation.
+#
+#   This library is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+# details.
+#
+#   You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+#++
+
+require "webrick"
+
+module Yast
+  # a webrick servlet which dynamically creates a tarball
+  # with the content of the current Git checkout
+  class TarballServlet < WEBrick::HTTPServlet::AbstractServlet
+    def do_GET(_request, response)
+      response.status = 200
+      response.content_type = "application/gzip"
+      response.body = source_archive
+    end
+
+  private
+
+    # compress the current sources into a tarball,
+    # no caching to ensure we always provide the latest content
+    def source_archive
+      # pack all Git files (including the non-tracked files (-o),
```

Comment:
Which files are included deserves a little discussion, and a section in the README.

Compare, `rake osc:build` does not tolerate untracked files. You have to either remove, .gitignore, or git-add them. It is a bit of a pain but it makes me clean up, and it fails early. I like that, and would expect `rake server` to behave the same way. Or just say why you want it to behave differently.

Reply:
My use case is:

- There is a bug reported for the installer
- I find a suspicious place in the code
- I want to add some debugging calls or try a fix and see if that helps
- I do not want to commit or add anything into Git, the debugging call is just a temporary hack or the proposed fix might not help in the end, I do not want to keep the changes
- I keep the changed file open in the editor, if I need to update the file I just save the changes and run `yupdate` in the inst-sys again, no need to bother with Git...
- I want to simply try something else or just run `git reset --hard` if it was completely wrong and start again

If the proposed fix works then the check for the untracked/modified files will be executed later in `rake osc:build` anyway. That `rake osc:build` check is sometimes quite annoying, having it once is enough for me.

Reply:
I see. But then you are simply packing all files and calling Git is only a complication. Why don't you

```sh
find . -name .git -prune -o -print
```

compare:
```sh
diff -u <(git ls-files --cached --others | sort) <(find . -name .git -prune -o -print | sed 's@^./@@' | sort) | less
```

Reply:
Well, we could do the file listing without git, the question is how to skip the `.gitignore`d files? 

Some ignored files are quite big and are not need at all, e.g. `package/*.tar.bz2` which already contains the complete sources. I'd like to skip that files.

Reply:
Ah, there should be `--exclude-standard` in my `git ls-files` approach...

Reply:
Nice, now it makes more sense :)